src/src/Commands/GPIO.cpp

```diff
@@ -0,0 +1,392 @@
+      mode=PIN_MODE_OUTPUT;
+      state=(event->Par2==0)?0:1;
+    }
+
+    const uint32_t key = createKey(pluginID,event->Par1);
+
+    if (globalMapPortStatus[key].mode != PIN_MODE_OFFLINE)
+    {
+      int8_t currentState;
+      GPIO_Read(pluginID, event->Par1, &currentState);
+      if (currentState==-1) {
+        mode=PIN_MODE_OFFLINE;
+        state = -1;
+      }
+
+      createAndSetPortStatus_Mode_State(key,mode,state);
+
+      switch (pluginID) {
+        case PLUGIN_GPIO:
+          if (mode==PIN_MODE_OUTPUT) GPIO_Internal_Write(event->Par1, state);
+  				break;
+  			case PLUGIN_MCP:
+  				GPIO_MCP_Write(event->Par1, state);
+  				break;
+  			case PLUGIN_PCF:
+  				GPIO_PCF_Write(event->Par1, state);
+  				break;
+      }
+  		String log = logPrefix + String(F(" : port#")) + String(event->Par1) + String(F(": set to ")) + String(state);
+  		addLog(LOG_LEVEL_INFO, log);
+  		SendStatusOnlyIfNeeded(event->Source, SEARCH_PIN_STATE, key, log, 0);
+  		return return_command_success();
+  	} else {
+      logErrorGpioOffline(logPrefix,event->Par1);
+      return return_command_failed();
+    }
+  } else {
+    logErrorGpioOutOfRange(logPrefix,event->Par1);
+    return return_command_failed();
+  }
+}
+
+void logErrorGpioOffline(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(": port# ")) + String(port) + String(F(" is offline."));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void logErrorGpioOutOfRange(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(" port#")) + String(port) + String(F(" is out of range"));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void logErrorGpioNotOutput(String prefix, byte port)
+{
+  String log;
+  log = prefix + String(F(" port#")) + String(port) + String(F(" is not an output port"));
+  addLog(LOG_LEVEL_ERROR, log);
+}
+
+void createAndSetPortStatus_Mode_State(uint32_t key, byte newMode, int8_t newState)
+{
+  // WARNING: operator [] creates an entry in the map if key does not exist
+
+  globalMapPortStatus[key].mode     = newMode;
+  globalMapPortStatus[key].command  = 1; //set to 1 in order to display the status in the PinStatus page
+
+  //only force events if state has changed
+  if (globalMapPortStatus[key].state != newState) {
+    globalMapPortStatus[key].state        = newState;
+    globalMapPortStatus[key].forceEvent   = 1;
+    globalMapPortStatus[key].forceMonitor = 1;
+  }
+}
+
+bool getPluginIDAndPrefix(char selection, byte* pluginID, String* logPrefix)
+{
+  bool success = true;
+  switch(selection)
+  {
+    case 'g': //gpio
+      *pluginID=PLUGIN_GPIO;
+      *logPrefix=F("GPIO");
+      break;
+    case 'm': //mcp
+      *pluginID=PLUGIN_MCP;
+      *logPrefix=F("MCP");
+      break;
+    case 'p': //pcf
+      *pluginID=PLUGIN_PCF;
+      *logPrefix=F("PCF");
+      break;
+    default:
+      success=false;
+  }
+  return success;
+}
+
+bool getPluginIDAndPrefixAndType(char selection, byte* pluginID, String* logPrefix, byte* gpioTimerType)
```

Comment:
Why use pointers here for the parameters?
They are not optional, right? So better use `&` here.

Reply:
you are right

Reply:
done